firefox-ios/Client/Frontend/Home/TopSites/Cell/ObservableCollectionViewCell.swift

```diff
@@ -0,0 +1,167 @@
+    var isVisibilityMonitoringEnabled = false {
+        didSet {
+            if isVisibilityMonitoringEnabled {
+                startObservingIfNeeded()
+                checkIfCellIsInView()
+            } else {
+                stopObserving()
+                stopVisibilityTimer()
+                wasPreviouslyInView = false
+            }
+        }
+    }
+    var inViewFractionThreshold: CGFloat = 0.5
+    var onAboveInViewThreshold: ((ObservableCollectionViewCell) -> Void)?
+    var visibleTimeThresholdSeconds: TimeInterval = 1.0
+    var onAboveVisibleTimeThreshold: ((ObservableCollectionViewCell) -> Void)?
+
+    // MARK: In-view State
+    var isInView: Bool { inViewAreaFraction >= inViewFractionThreshold }
+    private var wasPreviouslyInView = false
+
+    // MARK: Visibility State
+    private var visibilityTimer: Timer?
+    private var wasVisibleForThisLifetime = false
+
+    // MARK: In-view Fraction Logic
+    private var observedScrollViews: Set<UIScrollView> = []
+
+    private var inViewAreaFraction: CGFloat {
+        guard let window = window, !isHidden, alpha > 0.01, !bounds.isEmpty else { return 0 }
+        var rect = convert(bounds, to: window).intersection(window.bounds)
+        if rect.isNull { return 0 }
+        var a = superview
+        while let s = a, s !== window {
+            if s.clipsToBounds {
+                rect = rect.intersection(s.convert(s.bounds, to: window))
+                if rect.isNull { return 0 }
+            }
+            a = s.superview
+        }
+        let total = bounds.width * bounds.height
+        guard total > 0 else { return 0 }
+        return max(0, min(1, (rect.width * rect.height) / total))
+    }
+
+    private var scrollViews: [UIScrollView] {
+        let chain = sequence(first: superview, next: { $0?.superview })
+        return Array(chain).compactMap { $0 as? UIScrollView }
+    }
+
+    // MARK: Lifecycle
+    override func prepareForReuse() {
+        stopObserving()
+        stopVisibilityTimer()
+        wasPreviouslyInView = false
+        wasVisibleForThisLifetime = false
+        super.prepareForReuse()
+    }
+
+    override func layoutSubviews() {
+        if isVisibilityMonitoringEnabled {
+            startObservingIfNeeded()
+            checkIfCellIsInView()
+        }
+        super.layoutSubviews()
+    }
+
+    override func willMove(toWindow newWindow: UIWindow?) {
+        // If the cell leaves the screen without prepareForReuse() being called we want to make sure we stop the timer
+        if newWindow == nil {
+            stopVisibilityTimer()
+        }
+        super.willMove(toWindow: newWindow)
+    }
+
+    // MARK: Observing Logic
+    private func startObservingIfNeeded() {
+        for sv in scrollViews where !observedScrollViews.contains(sv) && isVisibilityMonitoringEnabled {
+            sv.addObserver(self, forKeyPath: "contentOffset", context: nil)
+            observedScrollViews.insert(sv)
+        }
```

Comment:
I would think we would need to only observe the parent scrollview and not all scrollviews?